Triangulation_3/include/CGAL/IO/Triangulation_raw_iostream_3.h

```diff
@@ -0,0 +1,396 @@
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: LGPL-3.0-or-later OR LicenseRef-Commercial
+//
+// Author(s)     : Ange Clément
+
+
+#ifndef CGAL_TRIANGULATION_RAW_IOSTREAM_3_H
+#define CGAL_TRIANGULATION_RAW_IOSTREAM_3_H
+
+#include <CGAL/license/Triangulation_3.h>
+
+#include <iostream>
+#include <vector>
+
+#include <CGAL/Unique_hash_map.h>
+#include <CGAL/Default.h>
+#include <CGAL/IO/io.h>
+
+namespace CGAL {
+
+template < class GT, class Tds = Default,
+         class Lock_data_structure = Default >
+class Triangulation_3;
+
+namespace IO {
+
+namespace internal {
+
+template < class GT, class Tds, class Lds >
+bool construct_vertices_map(const Triangulation_3<GT, Tds, Lds>& tr, Unique_hash_map<typename Tds::Vertex_handle, std::size_t>& vertices_map)
+{
+  typedef Triangulation_3<GT, Tds>                 Triangulation;
+  typedef typename Triangulation::size_type        size_type;
+  typedef typename Triangulation::Vertex_iterator  Vertex_iterator;
+
+  size_type i = 0;
+  vertices_map[tr.infinite_vertex()] = 0;
+  for(Vertex_iterator it = tr.vertices_begin(), end = tr.vertices_end(); it != end; ++it)
+  {
+    vertices_map[it] = i++;
+  }
+
+  CGAL_assertion(i == tr.number_of_vertices()+1);
+
+  return true;
+}
+
+template < class GT, class Tds, class Lds >
+bool write_header(std::ostream& os, const Triangulation_3<GT, Tds, Lds>& tr)
+{
+  // Writes:
+  // - the dimension
+  if(IO::is_ascii(os))
+  {
+    os << tr.dimension() << "\n";
+  }
+  else
+  {
+    write(os, tr.dimension());
+  }
+  return (bool)os;
+}
+
+template < class GT, class Tds, class Lds >
+bool read_header(std::istream& is, Triangulation_3<GT, Tds, Lds>& tr)
+{
+  // Reads:
+  // - the dimension
+  int dimension;
+  if(IO::is_ascii(is))
+  {
+    is >> dimension;
+  }
+  else
+  {
+    read(is, dimension);
+  }
+  if(dimension > 3 || dimension < -2)
+    return false;
+  tr.tds().set_dimension(dimension);
+  return (bool)is;
+}
+
+template < class GT, class Tds, class Lds >
+bool write_vertices(std::ostream& os, const Triangulation_3<GT, Tds, Lds>& tr)
+{
+  // Writes:
+  // - the number of finite vertices
+  // - the non combinatorial information on vertices (point, etc)
+  typedef Triangulation_3<GT, Tds>                 Triangulation;
+  typedef typename Triangulation::size_type        size_type;
+  typedef typename Triangulation::Vertex_iterator  Vertex_iterator;
+
+  size_type n = tr.number_of_vertices();
+  if (n == 0)
+    return false;
+
+  if(IO::is_ascii(os))
+  {
+    os << n << '\n';
+  }
+  else
+  {
+    write(os, n);
+  }
+
+  for(Vertex_iterator it = ++tr.vertices_begin(), end = tr.vertices_end(); it != end; ++it)
+  {
+    os << *it;
+    if(IO::is_ascii(os))
+      os << '\n';
+  }
+
+  return (bool)os;
+}
+
+struct IdentityConvertVertex
+{
+  template <class V>
+  void operator()(const V& v_input, V& v_output)
+  {
+    auto time_stamp_ = v_output.time_stamp();
+    v_output = v_input;
+    v_output.set_time_stamp(time_stamp_);
+  }
+};
+
```

Comment:
This class can only work if the vertex has a timestamp. It will compile for Mesh_3, for not for Triangulation_3.